### DISCUSSION TITLE
Context injection deprecation

### DISCUSSION THREAD

**Discussion by jbrwn** [ExpediaGroup/graphql-kotlin#1305] [Q&A]
Im upgrading to v5.2.0 and I am curious what the long term plan is around the now deprecated context injection functionality. 

https://github.com/ExpediaGroup/graphql-kotlin/blob/master/website/docs/schema-generator/execution/contextual-data.md#interface-injection-deprecated

The example below will not generate any deprecation warnings even though the `GraphQLContext` interface itself has been deprecated.  This may sneak up on some users when the injection "magic" suddenly stops working.  Not sure what the best approach is here.  Maybe log a warning at runtime?  Thanks in advance for taking a look at this.

```class ContextualQuery : Query {
    fun contextualQuery(
        context: MyGraphQLContext,
        value: Int
    ): String = "The custom value was ${context.customValue} and the value was $value"
}
```

**Comment by smyrick** [comment]
If you look at the section just above where you linked, we have outlined why we deprecated the interface. This features is really just a shorthand for the `DataFetchingEnvironment` injection that is already supported by graphql-java

```kotlin
// Set in the context factory
class MyContext(val foo: String)

class ContextualQuery : Query {
    fun contextualQuery(dfe: DataFetchingEnvironment, value: Int): String {
        val context = dfe.context as MyContext
        return "The context value was ${context.foo} and the value was $value"
    }
}
```

You should migrate to this pattern going forward and stop using our custom interface.

--------

The second part is that in graphql-java 17, the context changed from a generic object that you create to being a `Map<*,Any>`. This means that you have to use the explicit get and set methods on the map, but then you can put any values or objects you want this map. So instead of `dfe.context` you should use `dfe.graphqlContext`.

**Reply by smyrick** [reply]
For more details on the context map change, see the initial PR: https://github.com/graphql-java/graphql-java/pull/2368

**Reply by jbrwn** [reply]
@smyrick appreciate the response. I understand the change.  I’m not sure my questions were addressed so I’ll be more direct: 

1. when will the deprecated code actually be removed from the code base?
2. can you add some kind of runtime warning for the injection pattern I described above? The only way to know context injection is deprecated is by looking at the docs or release notes. We have many engineers contributing to our graphql-kotlin based service. A compile or runtime warning would help us migrate away from this pattern.

**Reply by jbrwn** [reply]
In regards to item 2, I can possibly help implement if you are open adding the warning

**Reply by dariuszkuc** [reply]
Hello 👋 
Sadly we don't have a timeline for the releases so cannot say for sure when that deprecated functionality will be removed. My guess removal will be mostly driven by the changes in upstream (i.e. `graphql-java` dropping this support) or by potential migration to the Spring GraphQL starter (AFAIK they are targeting GA sometime in the Spring 22). Sorry cannot be more specific.

As for the warning log message -> that's a great idea. Feel free to open an issue to track this and we are always open to contributions.

**Reply by jbrwn** [reply]
@dariuszkuc @smyrick PR for runtime warning log message -> https://github.com/ExpediaGroup/graphql-kotlin/pull/1306.  Thanks again.

**Reply by floatdrop** [reply]
@smyrick still it would be more convenient to inject `GraphQLContext` as well as `DataFetchingEnvironment` – because it can be constructed for tests, but `DataFetchingEnvironment` can only be mocked.

**Reply by smyrick** [reply]
@floatdrop I am no longer a core-maintainer of this project but the context interface method is still deprecated, and in my opinion should now be removed. The new generic map method in graphql-java can still be updated to have a property that you expect to cast to a specific type but the graphql-kotlin library should leave that to users to handle and what to do incase casting does not work.

cc @samuelAndalon to consider dropping `GraphQLContext` interface in next major version

**Reply by dariuszkuc** [reply]
fyi https://github.com/ExpediaGroup/graphql-kotlin/issues/1548
